File: 21.01.20/Programmers_67256.py
def solution(numbers, hand):
    answer = ''

    arr = [(3, 1), (0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2), (3, 0), (-1, -1), (3, 2)]
    pos_left, pos_right = arr[10], arr[12]
    for i in range(len(numbers)):
        if numbers[i] in [1, 4, 7]:
            answer += 'L'
            pos_left = arr[numbers[i]]
        elif numbers[i] in [3, 6, 9]:
            answer += 'R'
            pos_right = arr[numbers[i]]
        else:
            temp, pos_left, pos_right = dis_calc(arr[numbers[i]], pos_left, pos_right, hand)
            answer += temp

    return answer


def dis_calc(target, pos_left, pos_right, hand):
    if target == pos_left:
        return 'L', target, pos_right
    elif target == pos_right:
        return 'R', pos_left, target

    # dis_l = (abs(pos_left[0] - target[0]) ** 2 + abs(pos_left[1] - target[1]) ** 2) ** 0.5
    # dis_r = (abs(pos_right[0] - target[0]) ** 2 + abs(pos_right[1] - target[1]) ** 2) ** 0.5
    dis_l = abs(pos_left[0] - target[0]) + abs(pos_left[1] - target[1])
    dis_r = abs(pos_right[0] - target[0]) + abs(pos_right[1] - target[1])

    if dis_l < dis_r:
        return 'L', target, pos_right
    elif dis_l > dis_r:
        return 'R', pos_left, target
    elif dis_l == dis_r:
        if hand == 'left':
            return 'L', target, pos_right
        else:
            return 'R', pos_left, target







print(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"))
print(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], "left"))
print(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0], "right"))
print(solution([2, 0, 5, 8, 2, 0, 8, 5, 2, 0, 5, 8], 'left'))

# answer
# "LRLLLRLLRRL"
# "LRLLRRLLLRR"
# "LLRLLRLLRL"

File: 21.04.17/Programmers_67256.js
function solution(numbers, hand) {
  var answer = '';

  // const arr = new Array()
  const dial = {
    1: [0, 0], 2: [0, 1], 3: [0, 2],
    4: [1, 0], 5: [1, 1], 6: [1, 2],
    7: [2, 0], 8: [2, 1], 9: [2, 2],
    '*': [3, 0], 0: [3, 1], '#': [3, 2],
  };
  let left_thumb = dial['*'];
  let right_thumb = dial['#'];
  for (var i = 0; i < numbers.length; i++) {
    if (numbers[i] === 1 | numbers[i] === 4 | numbers[i] === 7) {
      left_thumb = dial[numbers[i]]
      answer += 'L';
    } else if (numbers[i] === 3 | numbers[i] === 6 | numbers[i] === 9) {
      right_thumb = dial[numbers[i]]
      answer += 'R';
    } else {
      if (Math.abs(left_thumb[0] - dial[numbers[i]][0]) + (Math.abs(left_thumb[1] - dial[numbers[i]][1])) < (Math.abs(right_thumb[0] - dial[numbers[i]][0])) + (Math.abs(right_thumb[1] - dial[numbers[i]][1]))) {
        left_thumb = dial[numbers[i]]
        answer += 'L';
      } else if (Math.abs(left_thumb[0] - dial[numbers[i]][0]) + (Math.abs(left_thumb[1] - dial[numbers[i]][1])) > (Math.abs(right_thumb[0] - dial[numbers[i]][0])) + (Math.abs(right_thumb[1] - dial[numbers[i]][1]))) {
        right_thumb = dial[numbers[i]]
        answer += 'R';
      } else {
        if (hand === 'left') {
          left_thumb = dial[numbers[i]]
          answer += 'L';
        } else {
          right_thumb = dial[numbers[i]]
          answer += 'R';
        }
      }
    }
  }

  return answer;
}




console.log(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"))
console.log(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], "left"))
console.log(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0], "right"))

// answer
// "LRLLLRLLRRL"
// "LRLLRRLLLRR"
// "LLRLLRLLRL"